### ISSUE TITLE
Lifespan of processes inside `trainer.fit(devices=-1, accelerator="gpu")` in 2.0.x

### ISSUE THREAD

**Issue by jakub-h** [Lightning-AI/pytorch-lightning#18761]
### Bug description

Hi, I've experienced strange behavior of Lightning v2.0.x regarding multiprocessing recently. When I run `trainer.fit()` on multiple GPUs (`accelerator="gpu", devices=-1` using `CUDA_VISIBLE_DEVICES="0,1,2"`), I expect the parallel part will end with the end of `.fit()` method. Everything after the calling of `trainer.fit()` should be run in a single process, imho. It does so in `v1.9.x`, but not in `v2.0.x`. It seems that when you train e.g. on 3 devices in parallel (using "ddp" strategy), 3 processes continue to live even after the `fit()` method is finished and process the rest of the script simultaneously.

How to reproduce the bug:
Run following minimal code snippet (`CUDA_VISIBLE_DEVICES="0,1,2" snippet.py`) with lightning `1.9.0` and the "######### Something after the .fit() method." message is written only once compared to 3 times when using `2.0.0`.

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
import os

import torch
from lightning.pytorch import LightningModule, Trainer
from torch.utils.data import DataLoader, Dataset


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        devices=-1,
        num_nodes=1,
        accelerator="gpu",
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    print("######### Something after the .fit() method.")


if __name__ == "__main__":
    run()
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA TITAN X (Pascal)
                - NVIDIA GeForce GTX 1080 Ti
                - NVIDIA TITAN X (Pascal)
                - NVIDIA GeForce GTX 1080 Ti
                - NVIDIA GeForce GTX 1080 Ti
                - NVIDIA GeForce GTX 1080 Ti
                - NVIDIA GeForce GTX 1080 Ti
                - NVIDIA GeForce GTX 1080 Ti
                - NVIDIA GeForce GTX 1080 Ti
                - NVIDIA GeForce GTX 1080 Ti
        - available:         True
        - version:           12.1
* Lightning:
        - lightning:         2.0.0
        - lightning-cloud:   0.5.39
        - lightning-utilities: 0.9.0
        - pytorch-lightning: 2.0.9.post0
        - torch:             2.1.0
        - torchmetrics:      1.2.0
        - torchvision:       0.16.0
* Packages:
        - aiohttp:           3.8.5
        - aiosignal:         1.3.1
        - annotated-types:   0.5.0
        - anyio:             3.7.1
        - arrow:             1.3.0
        - async-timeout:     4.0.3
        - attrs:             23.1.0
        - backoff:           2.2.1
        - beautifulsoup4:    4.12.2
        - black:             23.9.1
        - blessed:           1.20.0
        - certifi:           2023.7.22
        - charset-normalizer: 3.3.0
        - click:             8.1.7
        - croniter:          1.3.15
        - dateutils:         0.6.12
        - deepdiff:          6.6.0
        - exceptiongroup:    1.1.3
        - fastapi:           0.88.0
        - filelock:          3.12.4
        - flake8:            6.1.0
        - frozenlist:        1.4.0
        - fsspec:            2023.9.2
        - h11:               0.14.0
        - idna:              3.4
        - inquirer:          3.1.3
        - isort:             5.12.0
        - itsdangerous:      2.1.2
        - jinja2:            3.1.2
        - lightning:         2.0.0
        - lightning-cloud:   0.5.39
        - lightning-utilities: 0.9.0
        - markdown-it-py:    3.0.0
        - markupsafe:        2.1.3
        - mccabe:            0.7.0
        - mdurl:             0.1.2
        - mpmath:            1.3.0
        - multidict:         6.0.4
        - mypy:              1.5.1
        - mypy-extensions:   1.0.0
        - networkx:          3.1
        - numpy:             1.26.0
        - nvidia-cublas-cu12: 12.1.3.1
        - nvidia-cuda-cupti-cu12: 12.1.105
        - nvidia-cuda-nvrtc-cu12: 12.1.105
        - nvidia-cuda-runtime-cu12: 12.1.105
        - nvidia-cudnn-cu12: 8.9.2.26
        - nvidia-cufft-cu12: 11.0.2.54
        - nvidia-curand-cu12: 10.3.2.106
        - nvidia-cusolver-cu12: 11.4.5.107
        - nvidia-cusparse-cu12: 12.1.0.106
        - nvidia-nccl-cu12:  2.18.1
        - nvidia-nvjitlink-cu12: 12.2.140
        - nvidia-nvtx-cu12:  12.1.105
        - ordered-set:       4.1.0
        - packaging:         22.0
        - pathspec:          0.11.2
        - pillow:            10.0.1
        - pip:               22.0.2
        - platformdirs:      3.11.0
        - psutil:            5.9.5
        - pycodestyle:       2.11.0
        - pydantic:          1.10.13
        - pydantic-core:     2.4.0
        - pyflakes:          3.1.0
        - pygments:          2.16.1
        - pyjwt:             2.8.0
        - python-dateutil:   2.8.2
        - python-editor:     1.0.4
        - python-multipart:  0.0.6
        - pytorch-lightning: 2.0.9.post0
        - pytz:              2023.3.post1
        - pyyaml:            6.0.1
        - readchar:          4.0.5
        - requests:          2.31.0
        - rich:              13.6.0
        - setuptools:        59.6.0
        - six:               1.16.0
        - sniffio:           1.3.0
        - soupsieve:         2.5
        - starlette:         0.22.0
        - starsessions:      1.3.0
        - sympy:             1.12
        - tomli:             2.0.1
        - torch:             2.1.0
        - torchmetrics:      1.2.0
        - torchvision:       0.16.0
        - tqdm:              4.66.1
        - traitlets:         5.11.2
        - triton:            2.1.0
        - types-python-dateutil: 2.8.19.14
        - typing-extensions: 4.8.0
        - urllib3:           2.0.6
        - uvicorn:           0.23.2
        - wcwidth:           0.2.8
        - websocket-client:  1.6.3
        - websockets:        11.0.3
        - yarl:              1.9.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.10.6
        - release:           5.15.0-78-generic
        - version:           #85-Ubuntu SMP Fri Jul 7 15:25:09 UTC 2023

</details>

**Comment by awaelchli** [comment]
@jakub-h This is expected since with 2.0, the default is `strategy="ddp"` which launches subprocesses. If you want the old default behavior from 1.9, set `strategy="ddp_spawn"`.

**Comment by awaelchli** [comment]
Hey, sorry for not being available.

Yes, with "It is also strongly recommended to use a single device for testing and evaluating models" it is meant that you would run the testing in a separate script with a single device. 

The way you show it in the snippet wouldn't work well because the processes have already been created. Sorry for this limitation.  

My personal recommendation is:
1. One script to train the model with DDP
2. One script to test the model with a single device (loading the checkpoint from the training script)

This is the sane way to do testing in my personal opinion and experience.
